### PR TITLE
Support generating MSVC bindings

### PR DESCRIPTION
**Pull request by vvuk**
This is a WIP PR -- I'd like some feedback and suggestions on how to handle the msvc case properly (@emilio @bholley @nox).

I'm not familiar with libllvm/clang -- I need access to the currently configured target triple (x86_64-pc-win32 for the msvc case).  I see llvm::sys::getDefaultTargetTriple(), but I'm not sure if that will pick up the triple from a --triple argument.

But does this look sane in general?

### CODE REVIEW

**Comment by vvuk** [comment]
Also note that clang's MSVC support only works with up to Visual Studio 2013 (VC 12.0); it doesn't quite handle the 2015 headers properly.  Which is annoying, because 2013's c++11's support is still a bit wonky (e.g. hacky char16_t), which requires some hacks to get proper SM bindings generated.

File: src/gen.rs [hunk 1]
```diff
@@ -80,8 +80,16 @@ fn empty_generics() -> ast::Generics {
...
+        name.contains("@") ||
+        name.contains("?") ||
+        name.contains("$") ||
+        "bool" == name
+    {
         let mut s = name.to_owned();
+        s = s.replace("@", "_");
+        s = s.replace("?", "_");
+        s = s.replace("$", "_");
         s.push_str("_");
```
Comment (emilio) [inline src/gen.rs:83]: Potentially we just want to append the underscore if it's a rust keyword, right? This should also work though.

File: src/gen.rs [hunk 2]
```diff
@@ -225,7 +235,7 @@ fn gen_unmangle_method(ctx: &mut GenCtx,
                         span: ctx.span,
                         global: false,
                         segments: vec!(ast::PathSegment {
-                            identifier: ctx.ext_cx.ident_of(&v.mangled),
+                            identifier: ctx.ext_cx.ident_of(&mangled_rs),
```
Comment (emilio) [inline src/gen.rs]: Not sure if the generated function already does `rust_id(xxx)`, it seems plausible (I'll test later). If not, it should also be used there.

**Comment by emilio** [comment]
Related to getting the target triple, the only related thing I found in libclang is the `clang_getClangVersion` function, and I don't know if the triple would be contained there, plus it isn't a stable format per the docs.

Another option is to look for compiler-specific macro names in the parse stage. That's probably easy and reliable enough (at the end is what C/C++ programs do).

**Comment by vvuk** [comment]
Hmm, yeah, I guess I can look for _MSC_VER... that should work.  A few more fixes were pushed to the branch; associated JS trees and tweaks are mozjs https://github.com/vvuk/mozjs/tree/rust-msvc and rust-mozjs https://github.com/vvuk/rust-mozjs/tree/rust-msvc

**Comment by vvuk** [comment]
Going to close this and create a PR for rust-bindgen

**Pull request closed without merging**